services/auth/source/oauth2/init.go

```diff
@@ -71,12 +81,7 @@ func initOAuth2LoginSources() error {
 		}
 		err := oauth2Source.RegisterSource()
 		if err != nil {
-			log.Critical("Unable to register source: %s due to Error: %v. This source will be disabled.", source.Name, err)
-			source.IsActive = false
-			if err = models.UpdateSource(source); err != nil {
-				log.Critical("Unable to update source %s to disable it. Error: %v", err)
-				return err
-			}
+			log.Critical("Unable to register source: %s due to Error: %v.", source.Name, err)
```

Comment:
These codes added here to allow Gitea start up when some auth sources take down temporarily.

Reply:
Yes. 

The proposed changes in the Auth and Callback handlers allow us to stop auto disabling and reattempt to register at login time if registration failed at startup. 

(But you've reminded me I need to recheck GetProviders once we have the hard lock to prevent double registration.)

If that "second" attempt at registration fails again then the user will be presented with a 500 page - but it genuinely is an internal server error at that point - although I fully expect that there will be complaints about that too.

(Hopefully such complaints would actually come with logs so we could at least attempt to present a nicer error page.)

Reply:
> (But you've reminded me I need to recheck GetProviders once we have the hard lock to prevent double registration.)

Actually we're ok from the double register PoV but the issue is that RegisterSource could be somewhat slow so users may double/triple attempt to login. Will have a think.

Reply:
Hmm... looking at routers/web/user/auth.go:582 and the changes made by #14116 resetting is actually already handled.

So I'll let #14116 handle the re-registration attempt.